DOC-12351 Add or Remove Services and Rebalance (#3846)
couchbase/docs-server: modules/learn/pages/services-and-indexes/services/services.adoc

= Services
:description: pass:q[Couchbase Server Services provide data access and maintenance. You can deploy services flexibly across available hardware to support Multi-Dimensional Scaling, which allows the tuning of the cluster for optimal performance as workloads change.]
:page-toclevels: 3
:page-aliases: understanding-couchbase:services-and-indexes/services/services,architecture:services-archi-multi-dimensional-scaling,clustersetup:services-mds,learn:services-and-indexes/services-and-indexes,understanding-couchbase:services-and-indexes/services-and-indexes

[abstract]
{description}

> Services in Couchbase Server are configured and deployed during initialization on one or more nodes by a Full Administrator.

Reviewer: ```suggestion
Services in Couchbase Server are configured and deployed during initialization on 1 or more nodes by a Full Administrator.
```

During configuration, the administrator can select a subset of services for each node and assign memory to each service individually.

Each service provides a specific type of data access.

> Deploy only required services.

Reviewer: ```suggestion
Couchbase recommends deploying only required services.
```

To handle heavy workloads, you can deploy a service across multiple nodes in the cluster to improve performance and resource availability.

> In Couchbase Server 7.6 and later versions, you can add one or more arbiter nodes to your cluster.

Reviewer: ```suggestion
In Couchbase Server 7.6 and later versions, you can add 1 or more arbiter nodes to your cluster.
```

include::learn:partial$arbiter-node-benefits.adoc[]

== Services and Multi-Dimensional Scaling

Couchbase Server provides the following services:

* xref:learn:services-and-indexes:services/data-service.adoc[Data Service]: Stores, updates, and retrieves data items by key.

> * xref:learn:services-and-indexes:services/query-service.adoc[Query Service]: Parses and executes N1QL queries, and returns results.

Reviewer: ```suggestion
* xref:learn:services-and-indexes:services/query-service.adoc[Query Service]: Parses and executes {sqlpp} queries, and returns results.
```

It's not N1QL anymore

Interacts with the Data and Index services.
* xref:learn:services-and-indexes:services/index-service.adoc[Index Service]: Creates indexes, for use by the Query and Analytics services.
* xref:learn:services-and-indexes:services/search-service.adoc[Search Service]: Creates indexes for full-text search.

> Supports language-aware search, allowing queries for `beauties` to also match `beauty` and `beautiful`.

Reviewer: ```suggestion
* xref:learn:services-and-indexes:services/search-service.adoc[Search Service]: Creates indexes for a fully customizable search experience.
Supports near real-time search capabilities for a diverse range of data types, such as structured or unstructured text, dates, numbers, CIDR notation, geospatial data, and vectors.
```

* xref:learn:services-and-indexes:services/analytics-service.adoc[Analytics Service]: Performs join, set, aggregation, and grouping operations.
Designed for large, long-running workloads that require significant memory and CPU resources.
* xref:learn:services-and-indexes:services/eventing-service.adoc[Eventing Service]: Handles near real-time responses to data changes.
Executes code in response to document mutations or on scheduled timers.
* xref:learn:services-and-indexes:services/backup-service.adoc[Backup Service]: Schedules full and incremental backups for specific buckets or all buckets in a cluster.
Supports merging and pruning of existing backups.

With Multi-Dimensional Scaling (MDS), you can deploy, maintain, and provision these services independently to optimize performance as business needs and workloads change.

[#setting-up-services]
== Setting Up Services

Services are configured per node.

> Each node can run one instance of a service and up to seven services in total.

Reviewer: ```suggestion
Each node can run 1 instance of a service and up to 7 services in total.
```

In Couchbase Enterprise Server 7.6 and later, you can configure a node to run no services.

> The Data Service must run on at least one node in the cluster.

Reviewer: ```suggestion
The Data Service must run on at least 1 node in the cluster.
```

> Some services are interdependent and require at least one instance of each dependency to be running.

Reviewer: ```suggestion
Some services are interdependent and require at least 1 instance of each dependency to be running.
```

For example, the Query Service depends on the Index and Data services.

When you initialize the first node in a cluster, that node's service configuration becomes the default for nodes added later.
You can change the default configuration for individual nodes by removing services, adding services, or both.

In Couchbase Server 8.0 and later versions, you can add or remove the following non-Data Services on-demand on existing nodes in a cluster, without adding or removing nodes.

* index (Index Service).
* n1ql (Query Service).
* fts (Search Service).
* cbas (Analytics Service).
* eventing (Eventing Service).
* backup (Backup Service).

Then a rebalance operation is automatically triggered to distribute the service workload across the nodes and complete the modification.

> For information about adding or removing non-Data Services on an existing node, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

Reviewer: ```suggestion
For more information about adding or removing non-Data Services on an existing node, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].
```

> Adding or removing of kv (Data Service) on an existing node is supported only through adding or removing nodes.

Reviewer: ```suggestion
Adding or removing of the Data Service on an existing node is supported only through adding or removing nodes.
```

> For information about adding or removing a Data Service on an existing node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Manage a Data Service and Rebalance].

Reviewer 2: For information about adding or removing a Data Service on an existing node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Adding or Removing the Data Service on existing nodes].

> For information about adding or removing a Data Service on an existing node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Adding or Removing the Data Service on Existing Nodes].

Reviewer: ```suggestion
For more information about adding or removing the Data Service on an existing node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Adding or Removing the Data Service on Existing Nodes].
```

In Couchbase Server 8.0 and later versions, you can add or remove the following non-Data Services on-demand on existing nodes in a cluster, without adding or removing nodes.

* index (Index Service).
* n1ql (Query Service).
* fts (Search Service).
* cbas (Analytics Service).
* eventing (Eventing Service).
* backup (Backup Service).

Then a rebalance operation is automatically triggered to distribute the service workload across the nodes and complete the modification.
For information about adding or removing non-Data Services on an existing node, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

Adding or removing of kv (Data Service) on an existing node is supported only through adding or removing nodes.
For information about adding or removing a Data Service on an existing node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Adding or Removing the Data Service on Existing Nodes].

When the first node in a cluster is initialized, the services assigned to it become the default assignment for each other node subsequently to be added to the cluster.

> However, this default can be departed from, node by node, with one or more services omitted from the default, and one or more added.

Reviewer: Why did we delete this? Is this no longer the case?

Author: No, that content still exists on this page. It's just rewritten for proper readability.
Check line 49:
"When you initialize the first node in a cluster, that node's service configuration becomes the default for nodes added later....."

> Design the service allocation based on workload analysis.

Reviewer: ```suggestion
You should design your service allocation based on workload analysis.
```

> The following example shows how four services such as Data, Index, Query, and Search that can be evenly allocated across five nodes in a development cluster.

Reviewer: ```suggestion
The following example shows how 4 services such as Data, Index, Query, and Search that can be evenly allocated across 5 nodes in a development cluster:
```

> However, if large volumes of data require intensive indexing and querying with Search, the following production configuration is more efficient.

Reviewer: ```suggestion
However, if large volumes of data require intensive indexing and querying with Search, the following production configuration is more efficient:
```

> In the revised configuration, the Data Service runs exclusively on three nodes, the Index Service runs on two nodes, and the Query and Search services share the sixth node.

Reviewer: ```suggestion
In the revised configuration, the Data Service runs exclusively on 3 nodes, the Index Service runs on 2 nodes, and the Query and Search services share the sixth node.
```

> * Adding or removing a Data Service on a node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Manage a Data Service and Rebalance].

Reviewer 2: * Adding or removing a Data Service on a node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Adding or Removing the Data Service on existing nodes].

> * Adding or removing a Data Service on a node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Adding or Removing the Data Service on Existing Nodes].

Reviewer: ```suggestion
* Adding or removing the Data Service on a node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Adding or Removing the Data Service on Existing Nodes].
```

> * Using REST API to assign services to a new single node cluster, see xref:rest-api:rest-set-up-services.adoc[Assigning Services to a New Single Node].

Reviewer: ```suggestion
* Using the REST API to assign services to a new single node cluster, see xref:rest-api:rest-set-up-services.adoc[Assigning Services to a New Single Node].
```

> * Using REST API to assign services to an existing node of a cluster, see xref:rest-api:rest-set-up-services-existing-nodes.adoc[Assigning Services to an Existing Node].

Reviewer: ```suggestion
* Using the REST API to assign services to an existing node of a cluster, see xref:rest-api:rest-set-up-services-existing-nodes.adoc[Assigning Services to an Existing Node].
```

[#multi-dimensional-scaling]
== Multi-Dimensional Scaling

Couchbase Services can be deployed flexibly across hardware resources to support Multi-Dimensional Scaling.
This allows you to fine-tune a cluster for optimal performance as workloads change.

> For example, if Search workloads increase, you can remove other services from one or more non-Search nodes and reconfigure those nodes to run the Search Service in the cluster.

Reviewer: ```suggestion
For example, if Search workloads increase, you can remove other services from 1 or more non-Search nodes and reconfigure those nodes to run the Search Service in the cluster.
```

[#setting-up-services]
== Setting Up Services

Services are configured per node.
Each node can run one instance of a service and up to seven services in total.
In Couchbase Enterprise Server 7.6 and later, you can configure a node to run no services.

The Data Service must run on at least one node in the cluster.
Some services are interdependent and require at least one instance of each dependency to be running.
For example, the Query Service depends on the Index and Data services.

When you initialize the first node in a cluster, that node's service configuration becomes the default for nodes added later.
You can change the default configuration for individual nodes by removing services, adding services, or both.

In Couchbase Server 8.0 and later versions, you can add or remove the following non-Data Services on-demand on existing nodes in a cluster, without adding or removing nodes.

* index (Index Service).
* n1ql (Query Service).
* fts (Search Service).
* cbas (Analytics Service).
* eventing (Eventing Service).
* backup (Backup Service).

> Then a rebalance operation is automatically triggered to evenly distribute the service workload across the nodes.

Reviewer 2: Omit "evenly".

Author: Done. Also, added "and complete the modificatio" as suggested in 'What's New' page.

For information about adding or removing non-Data Services on an existing node, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

Adding or removing of kv (Data Service) on an existing node is supported only through adding or removing nodes.

To add a Data Service on a node from the UI:

* xref:manage:manage-nodes/add-node-and-rebalance.adoc[Add a New Node and Rebalance] along with enabling the Data Service.

* xref:manage:manage-nodes/remove-node-and-rebalance.adoc[Remove a Node and Rebalance].

> Then xref:manage:manage-nodes/add-node-and-rebalance.html[add the node] back to the cluster along with enabling the Data Service.

Reviewer 2: This is a comment for lines 68 through 71.

If I want to add the Data service on an existing node that does not currently have the Data service, all I need to do is lines 70 and 71 -- remove the node, then add the node back to the cluster along with enabling the Data Service.

So, if you are trying to say that there are 2 ways to do this -- add a new node with a data service (line 68)   OR remove a node (that does not have a data service), then, add it back in along with enabling the data service (lines 70 and 71), you need an OR between the two.

Either fix the confusion by removing line 68 or fix the confusion by adding an OR on line 69.

To remove a Data Service from an existing node using the UI:

* xref:manage:manage-nodes/remove-node-and-rebalance.adoc[Remove a Node and Rebalance].

> Then xref:manage:manage-nodes/add-node-and-rebalance.html[add the node] back to the cluster by disabling the Data Service.

Reviewer 2: This is a comment about lines 66 to 76 which is super confusing.  
I think that it should be:

To add a Data Service on an existing node of a cluster from the UI:
Remove the node that you want to add the Data Service to using Remove a Node and Rebalance.  Then add the node back to the cluster along with enabling the Data Service.  (Alternatively, you can add a new node with the services configuration desired, including the Data Service, then remove the extraneous node that you no longer need.)

To remove a Data Service from an existing node of a cluster using the UI:
Remove the node with the Data Service that you want to remove using Remove a Node and Rebalance.  Then add the node back to the cluster along with disabling the Data Service.

During the node removal operation, if there are other services on the node (that you are removing for the add/remove Data Service reconfiguration), and you do not want those services to be impacted when the node is removed from the cluster, you may need to add those services to another node temporarily or add a new node temporarily with those services enabled.

The following example shows how four services such as Data, Index, Query, and Search that can be evenly allocated across five nodes in a development cluster.

[#cb_cluster_with_services_development]
include::partial$cluster-example-diagrams.adoc[tag="five-cluster-setup"]

This configuration may provide adequate performance for development and testing.
However, if large volumes of data require intensive indexing and querying with Search, the following production configuration is more efficient.

[#cb_cluster_with_services_production]
include::partial$cluster-example-diagrams.adoc[tag="six-cluster-setup"]

In the revised configuration, the Data Service runs exclusively on three nodes, the Index Service runs on two nodes, and the Query and Search services share the sixth node.

For more information about:

* Service memory quotas, see xref:buckets-memory-and-storage/memory.adoc[Memory].

* Initializing a cluster, see xref:manage:manage-nodes/create-cluster.adoc[Create a Cluster].

* Adding or removing services on a node, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

> * Assigning services to a new node, see xref:rest-api:rest-set-up-services.adoc[Assigning Services to a New Node].

Reviewer 2: `* Using REST API to assign services to a new single node cluster, see xref:rest-api:rest-set-up-services.adoc[Assigning Services to a New Single Node].`

> * Assigning services to an existing node, see xref:rest-api:rest-set-up-services-existing-nodes.adoc[Assigning Services to an Existing Node].

Reviewer 2: `* Using REST API to assign services to an existing node of a cluster, see xref:rest-api:rest-set-up-services-existing-nodes.adoc[Assigning Services to an Existing Node].`

[#setting-up-services]
== Setting Up Services

Services are configured per node.
Each node can run one instance of a service and up to seven services in total.
In Couchbase Enterprise Server 7.6 and later, you can configure a node to run no services.

The Data Service must run on at least one node in the cluster.
Some services are interdependent and require at least one instance of each dependency to be running.
For example, the Query Service depends on the Index and Data services.

When you initialize the first node in a cluster, that node's service configuration becomes the default for nodes added later.
You can change the default configuration for individual nodes by removing services, adding services, or both.

In Couchbase Server 8.0 and later versions, you can add or remove the following non-Data Services on-demand on existing nodes in a cluster, without adding or removing nodes.

* index (Index Service).
* n1ql (Query Service).
* fts (Search Service).
* cbas (Analytics Service).
* eventing (Eventing Service).
* backup (Backup Service).

> Then a rebalance operation is automatically triggered to distribute the service workload across the nodes and complete the modification.

Reviewer 2: Remove line 61.  It's too much detail for a general overview about services.

Author: I'll retain the sentence or else it will be an incomplete information about the Services.

For information about adding or removing non-Data Services on an existing node, see xref:manage:manage-nodes/modify-services-and-rebalance.adoc[Modify Services and Rebalance].

Adding or removing of kv (Data Service) on an existing node is supported only through adding or removing nodes.

> 

Reviewer 2: For information about the Data Service, see xref:learn:services-and-indexes:services/data-service.adoc[Data Service].

Author: Instead I've added the following content:
For information about adding or removing a Data Service on an existing node, see xref:manage:manage-nodes/manage-data-service-and-rebalance.adoc[Manage a Data Service and Rebalance].

To add a Data Service on an existing node of a cluster from the UI:

. Remove the node that you want to add the Data Service to using xref:manage:manage-nodes/remove-node-and-rebalance.adoc[Remove a Node and Rebalance].
. Then xref:manage:manage-nodes/add-node-and-rebalance.html[add the node] back to the cluster along with enabling the Data Service.
+
Alternatively, you can add a new node with the necessary services configuration, including the Data Service, using xref:manage:manage-nodes/add-node-and-rebalance.adoc[Add a New Node and Rebalance].

> Then remove the extraneous node that you no longer need using xref:manage:manage-nodes/remove-node-and-rebalance.adoc[Remove a Node and Rebalance].

Reviewer 2: Remove all the info about how to add/remove the Data Service.  This level of detail is not needed on this page.  The details are making the information on the page less clear.

Remove Lines 66 - 80

> To remove a Data Service from an existing node of a cluster using the UI:

Reviewer 2: Remove line 74.

. Remove the node with the Data Service that you want to remove using xref:manage:manage-nodes/remove-node-and-rebalance.adoc[Remove a Node and Rebalance].
. Then xref:manage:manage-nodes/add-node-and-rebalance.html[add the node] back to the cluster along with disabling the Data Service.
+
During the node removal operation, if there are other services on the node that you are removing for the Data Service reconfiguration, and you do not want those services to be impacted when the node is removed from the cluster, you may need to add those services to another node temporarily.

> Then add a new node temporarily with those services enabled.

Reviewer 2: Remove lines 76 through 80.
